wordpress-mobile/WordPress-Android
# Create sites in user's timezone

# Create sites in user's timezone

Fixes [#14400 Site timezone defaults to UTC instead of user timezone](https://github.com/wordpress-mobile/WordPress-Android/issues/14400)

**Related:**
- Depends on [WordPress-FluxC-Android PR #2268](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2268)

Sites created via the WordPress Android app given the UTC timezone regardless of the time zone of the device.
This caused an issue for users in timezones having a negative hour offset relative to the UTC time resulting issues where a scheduled post publishes immediately after pressing SCHEDULE on the editor. Because the time zone of the site was actually UTC, the users are unaware they choose the UTC time in the time picker when scheduling a post, and that time could have been already in the past.

The time zone of the site is not displayed in the time picker during post scheduling in the app (also not when creating a new WP.com site) potentially leading to further confusion as to why the scheduled post publishes immediately while the user is expecting it to publish at a time relative to the current time shown in the status bar on the device.

On web, the WP Admin Dashboard displays the time zone is when scheduling a post.
<details>
<summary>WP Admin Dashboard - Post Scheduling</summary>

| <img width="1167" alt="before_4postSchedulingViaWpAdmin" src="https://user-images.githubusercontent.com/4588074/152517672-054379fb-4d6d-4cb4-a54a-aa4d87d2388b.png"> |
| --- |
| _On the web we avoid confusion by displaying the timezone of the site is when scheduling a post._ |
</details>

This PR proposes to add the timezone identifier returned by [java.util.TimeZone.getDefault().getID()](https://garygregory.wordpress.com/2013/06/18/what-are-the-java-timezone-ids/) to the `options.timezone_string` field in the payload of the call to the WordPress.com REST API endpoint for creating a new site. This solution is aligned with the current [WordPress-iOS approach](https://github.com/wordpress-mobile/WordPress-iOS/pull/17821) and is based on my understanding that the API accepts timezones in the [tz database format](https://en.wikipedia.org/wiki/Tz_database) which appears to match the format returned by the Java TimeZone util.

## To Test

1. Create a new site using the app
2. **Verify** that the timezone of the newly created site matches the device's timezone
   1. Go to My Site → Site Settings and check what's displayed under Timezone
3. **Verify** that post schedule as intuitively expected:
	1. Create a new post
	2. Schedule it to publish in a few minutes from now
		1. tap the three dots at top right of the screen
		2. go to Post Settings → Publish Date → Date and Time
		3. Choose current day then pick a time 2 minutes ahead of the time displayed in status bar of the device
		4. tap Back → Tap SCHEDULE at the top right
	3. **⚠️ Important:** Due to a known API issue you have to visit the site to trigger the publishing of scheduled posts. You can tap the site url under the Site Title on the My Site screen for this.
	   <details><summary>More...</summary>

	   > The author expects visitors to see the post after the scheduled time has passed, and this works as expected regardless of the existing issue. But confusion could arise if the user checks if the post is published via the app at My Site → Posts → PUBLISHED tab, because the post is still under the SCHEDULED tab.

	   The issue was originally reported for the WordPress-iOS repo [here](https://github.com/wordpress-mobile/WordPress-iOS/issues/10456)
	   </details>
   d. **Verify** That the post is published after the scheduled time has passed.
   (_it should be visible via to My Site → Posts → PUBLISHED tab._)
4. **Verify** in the WP Admin Dashboard using a web browser on a computer that the time zone of the site matches with the time zone of the device you used for testing:
	1.  Go to Settings → General → "Site timezone"

**Recommended**:  Repeat the testing steps after changing the time zone on the device, to test another timezone.
_The more time zones we test the better 🙇‍♂️_

## Previews

<details>
  <summary>App: My Site → Posts → Scheduled</summary>

  | Before | After |
  | --- | --- |
  | ![before_1postScheduled](https://user-images.githubusercontent.com/4588074/152518139-1d21922d-43b4-459f-8dcd-b4049895521a.png) | ![after_1postScheduled](https://user-images.githubusercontent.com/4588074/152518193-bd249db2-fa87-4738-96d6-d8cccd880dc9.png) |
</details>

<details>
  <summary>WP Admin Dashboard: Settings → Site timezone</summary>

  | Before | After |
  | --- | --- |
  | <img width="1101" alt="before_1siteTimeZoneInWpAdmin" src="https://user-images.githubusercontent.com/4588074/152518757-45387316-27e0-4b9f-8bc6-2e9277b0eded.png"> | <img width="1172" alt="after_3siteTimeZoneInWpAdmin" src="https://user-images.githubusercontent.com/4588074/152518788-33cb0bb7-6685-4efd-99a6-d3f9d06087c6.png"> |
</details>

## Regression Notes

#### Potential unintended areas of impact
- Site creation
- Publishing time for scheduled posts
- Compatibility issues with the web version of WordPress (Site timezone, Post publishing)

The above could be caused by issues in time zones that were not tested.

#### What I did to test those areas of impact (or what existing automated tests I relied on)

- I tested by creating a site and verifying that scheduling posts works as expected and it is intuitive. By intuitive I mean what I select in the time picker as the publishing time of a post is relative to the the time displayed on the device status bar at the top left corner.
	- I tested first time with the `Europe/Brussels` timezone
	   ✅ the result was as expected
	- Then I changed the timezone of the device to `America/New_York` and repeated the test 
       ✅ this also produced the expected result

#### What automated tests I added (or what prevented me from doing so)

- I added a unit test to verify that the timezone identifier is propagated to the site creation API request payload: 
  `CreateSiteUseCaseTest.verifyPropagatesTimeZoneId`
- I updated the existing site creation tests in the [WordPress-FluxC-Android PR #2268](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2268) project to ensure the `options.timezone_string` is included in the payload of the API request.

PR submission checklist:
